src/errors.rs

```diff
@@ -1,19 +1,18 @@
-use crate::lang::{
-    query, Positioned, Query, QueryPosition, Span, VALID_AGGREGATES, VALID_INLINE, VALID_OPERATORS,
-};
+use crate::lang::{query, Positioned, Query, VALID_AGGREGATES, VALID_INLINE, VALID_OPERATORS};
 use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
 use failure::Fail;
-use nom::types::CompleteStr;
-use nom::ErrorKind;
 use num_derive::FromPrimitive;
 use num_traits::FromPrimitive;
 use std::convert::From;
+use std::ops::Range;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use strsim::normalized_levenshtein;
 
 /// Container for the query string that can be used to parse and report errors.
-pub struct QueryContainer {
-    query: String,
-    reporter: Box<dyn ErrorReporter>,
+pub struct QueryContainer<'a> {
+    pub query: String,
+    pub reporter: Box<dyn ErrorReporter + 'a>,
```

Comment:
this  `'a` is just self referential I think? I'd probably just make it `+ 'static`, there's no actual lifetime on error reporter

Reply:
I think I had to add it for the unit test in the lang file.  In there, the error reporter stores strings in a vector and that vector needs to have a longer lifetime so that the output can be verified.